### ISSUE TITLE
"is not a constructor" should be detected in compile time as "used before its declaration."

### ISSUE THREAD

**Issue by qballer** [Microsoft/TypeScript#29220]
**TypeScript Version:**  3.3.0-dev.201xxxxx
This repository demonstrates that I run in to the issue in ts@next. 
https://github.com/qballer/tsbug

**Search Terms:**
constructor, lexical scope, run time error vs compile time
**Code**

```ts
const b = getB() 

function getB() {
    // will throw C is not a constructor in runtime 
    return new C() // ** 
}

// Will generate:  Class 'C' used before its declaration.
// const instanceOfC = new C()

class C {}
```

**Expected behavior:**
Should generate:  Class 'C' used before its declaration.

**Actual behavior:**
Explodes in runtime

**Playground Link:** 
[Playground](http://www.typescriptlang.org/play/#src=const%20b%20%3D%20getB()%20%0D%0A%0D%0Afunction%20getB()%20%7B%0D%0A%20%20%20%20return%20new%20C()%0D%0A%7D%0D%0A%0D%0Aclass%20C%20%7B%7D)
[Repo](https://github.com/qballer/tsbug)

**Related Issues:**

**Comment by weswigham** [comment]
Riiiight, so generally we optimistically assume that within a function body, all of the things it captures will be appropriately initialized and not-TDZ'd because we don't track information that across function boundaries today. We've offhandedly mentioned the potential for tracking more effects in the future, but are wary of the performance implications of doing so.

**Comment by fatcerberus** [comment]
Indeed, that `getB()` would be completely kosher if it weren't called immediately from the global scope like you're doing (suppose it were an export, or was otherwise accessible from other parts of the program), so producingan compile-time error there would be unnecessarily restrictive.  You basically have three choices in cases like this:

1. Be pessimistic and break valid use cases
2. Be optimistic and miss some errors
3. Solve the halting problem

Option 3 is completely intractable so TypeScript generally goes with option 2 when faced with cases like this.

**Comment by weswigham** [comment]
@fatcerberus it's not quite so bleak an outlook as that - honestly it's just a matter of tracing the captured reads and writes of a function (which, ofc, includes the captures of any functions it calls...) and checking that those captures aren't TDZ'd/uninitialized at each callsite. It's just not an effect we track currently, as I said.

**Comment by fatcerberus** [comment]
Note though that TDZ is a *runtime* concept (hence the "temporal" part), it's not always statically determinable--something Babel had trouble with in the past--I don't know if they've solved it.

**Comment by qballer** [comment]
Not sure if detecting this is equivalent to the halting problem @fatcerberus, but like @weswigham said it has some performance cost. This code is also not the "greatest" in the world by this situation and many others can creep up on you. 

@weswigham did any benchmark run on such static analysis ?

**Comment by fatcerberus** [comment]
Certainly in simple cases like in the OP it’s easy to detect, but I suspect that detecting this *in the general case* 100% of the time is indeed equivalent to solving the halting problem.  I admit the possibility that I’m wrong about that; I’m certainly no compiler writer. :smile:

**Comment by weswigham** [comment]
> @weswigham did any benchmark run on such static analysis ?

Not really, no, it's just an intuition because it's an enhancement to what we already tracking during flow control which is already a tad expensive.

**Comment by weswigham** [comment]
> Any idea of difficulty to implement ?

Similar to control flow, probably, so a fair bit.
